toolchain/check/import_ref.cpp

```diff
@@ -67,21 +68,36 @@ class ImportRefResolver {
   auto Resolve(SemIR::InstId inst_id) -> SemIR::ConstantId {
     work_stack_.push_back({inst_id});
     while (!work_stack_.empty()) {
+      auto initial_work = work_stack_.size();
```

Comment:
If you're now tracking initial work here, and we're also tracking initial work in the helpers to determine whether retry is needed, is the finished indicator on ResolveResult needed?

Sorry, piecing this together as I think through approaches. Maybe std::optional<ConstantId> actually could work as I think through we ResolveResult couldn't be replaced without this.

Reply:
I think that's possible, and a resolution step should be asking for a retry if and only if it didn't add any more work. However, I think it'd be fragile to rely on that, and would make the resolve code less clear. I've added a `CHECK` that each resolve step says it's finished if and only if it added no new work, and updated the comment on the `finished` flag to point out that it's technically not necessary.

Reply:
Note, regarding clarity, initial_work should probably be declared closer to where it's used.

Reply:
> Why is it fragile?

It's a subtle rule for a resolver function to follow, for which we'd have no checking. Eg, if a resolver function adds a call to `GetLocalConstantId` in the wrong place, we have an infinite loop, or if it intends to retry but somehow doesn't add any work (as it might be reasonable to do after creating a forward declaration in some cases) then it wouldn't actually get a retry. It seems much less error-prone to me for a resolver function to need to explicitly say whether it wants to be called again or not, rather than inferring that from other state. I think it also makes the resolve functions more readable.

Reply:
> Note, regarding clarity, initial_work should probably be declared closer to where it's used. [The flow here is a bit hard to read now, maybe it could have reduced nesting with a `continue`]

Done.